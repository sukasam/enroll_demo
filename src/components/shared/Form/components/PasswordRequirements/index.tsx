import { useTranslate } from "Components/shared/Translate";
import { useEffect, useState } from "react";

interface PasswordCriteria {
    minLength: boolean;
    lowercase: boolean;
    uppercase: boolean;
    number: boolean;
    specialChar: boolean;
}

interface PasswordRequirementsProps {
    password: string;
    touchedFields: { password?: boolean };
}

function PasswordRequirements({
    password,
    touchedFields
}: PasswordRequirementsProps): JSX.Element {
    const translate = useTranslate();
    const [passwordCriteria, setPasswordCriteria] = useState<PasswordCriteria>({
        minLength: false,
        lowercase: false,
        uppercase: false,
        number: false,
        specialChar: false
    });

    useEffect(() => {
        const newPasswordCriteria: PasswordCriteria = {
            minLength: password?.length >= 8,
            lowercase: password ? /[a-z]/.test(password) : false,
            uppercase: password ? /[A-Z]/.test(password) : false,
            number: password ? /[0-9]/.test(password) : false,
            specialChar: password
                ? /[!@#$%^&*(),.?":{}|<>]/.test(password)
                : false
        };
        setPasswordCriteria(newPasswordCriteria);
    }, [password]);

    const getClassName = (criteria: boolean): string => {
        if (criteria) return "valid";
        return touchedFields.password ? "invalid" : "initial";
    };

    return (
        <div className="passwordRequirements">
            <div
                className="passwordLabel"
                data-testid="create_account_password_must"
            >
                {translate("create_account_password_must")}
            </div>
            <ul>
                <li
                    className={getClassName(passwordCriteria.minLength)}
                    data-testid="create_account_password_requirement_1"
                >
                    {translate("create_account_password_requirement_1")}
                </li>
                <li
                    className={getClassName(passwordCriteria.lowercase)}
                    data-testid="create_account_password_requirement_2"
                >
                    {translate("create_account_password_requirement_2")}
                </li>
                <li
                    className={getClassName(passwordCriteria.uppercase)}
                    data-testid="create_account_password_requirement_3"
                >
                    {translate("create_account_password_requirement_3")}
                </li>
                <li
                    className={getClassName(passwordCriteria.number)}
                    data-testid="create_account_password_requirement_4"
                >
                    {translate("create_account_password_requirement_4")}
                </li>
                <li
                    className={getClassName(passwordCriteria.specialChar)}
                    data-testid="create_account_password_requirement_5"
                >
                    {translate("create_account_password_requirement_5")}
                </li>
            </ul>
        </div>
    );
}

export default PasswordRequirements;
