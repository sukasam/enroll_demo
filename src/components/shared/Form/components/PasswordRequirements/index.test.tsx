import { render, screen } from "@testing-library/react";
import PasswordRequirements from "./index";

// Mock the useTranslate hook
jest.mock("Components/shared/Translate", () => ({
    useTranslate: (): ((key: string) => string) => (key: string) => key
}));

describe("PasswordRequirements", () => {
    const defaultProps = {
        password: "",
        touchedFields: { password: false }
    };

    it("renders correctly", () => {
        render(<PasswordRequirements {...defaultProps} />);
        expect(
            screen.getByTestId("create_account_password_must")
        ).toBeInTheDocument();
        expect(
            screen.getByTestId("create_account_password_requirement_1")
        ).toBeInTheDocument();
        expect(
            screen.getByTestId("create_account_password_requirement_2")
        ).toBeInTheDocument();
        expect(
            screen.getByTestId("create_account_password_requirement_3")
        ).toBeInTheDocument();
        expect(
            screen.getByTestId("create_account_password_requirement_4")
        ).toBeInTheDocument();
        expect(
            screen.getByTestId("create_account_password_requirement_5")
        ).toBeInTheDocument();
    });

    it("applies correct classes when password is empty and untouched", () => {
        render(<PasswordRequirements {...defaultProps} />);
        expect(
            screen.getByTestId("create_account_password_requirement_1")
        ).toHaveClass("initial");
        expect(
            screen.getByTestId("create_account_password_requirement_2")
        ).toHaveClass("initial");
        expect(
            screen.getByTestId("create_account_password_requirement_3")
        ).toHaveClass("initial");
        expect(
            screen.getByTestId("create_account_password_requirement_4")
        ).toHaveClass("initial");
        expect(
            screen.getByTestId("create_account_password_requirement_5")
        ).toHaveClass("initial");
    });

    it("applies correct classes when password is invalid and touched", () => {
        render(
            <PasswordRequirements
                password="weak"
                touchedFields={{ password: true }}
            />
        );
        expect(
            screen.getByTestId("create_account_password_requirement_1")
        ).toHaveClass("invalid");
        expect(
            screen.getByTestId("create_account_password_requirement_2")
        ).toHaveClass("valid");
        expect(
            screen.getByTestId("create_account_password_requirement_3")
        ).toHaveClass("invalid");
        expect(
            screen.getByTestId("create_account_password_requirement_4")
        ).toHaveClass("invalid");
        expect(
            screen.getByTestId("create_account_password_requirement_5")
        ).toHaveClass("invalid");
    });

    it("applies correct classes when password is valid", () => {
        render(
            <PasswordRequirements
                password="StrongP@ss1"
                touchedFields={{ password: true }}
            />
        );
        expect(
            screen.getByTestId("create_account_password_requirement_1")
        ).toHaveClass("valid");
        expect(
            screen.getByTestId("create_account_password_requirement_2")
        ).toHaveClass("valid");
        expect(
            screen.getByTestId("create_account_password_requirement_3")
        ).toHaveClass("valid");
        expect(
            screen.getByTestId("create_account_password_requirement_4")
        ).toHaveClass("valid");
        expect(
            screen.getByTestId("create_account_password_requirement_5")
        ).toHaveClass("valid");
    });

    it("updates classes when password changes", () => {
        const { rerender } = render(<PasswordRequirements {...defaultProps} />);
        expect(
            screen.getByTestId("create_account_password_requirement_1")
        ).toHaveClass("initial");

        rerender(
            <PasswordRequirements
                password="StrongP@ss1"
                touchedFields={{ password: true }}
            />
        );
        expect(
            screen.getByTestId("create_account_password_requirement_1")
        ).toHaveClass("valid");
    });

    // Snapshot test
    it("matches snapshot", () => {
        const { container } = render(
            <PasswordRequirements {...defaultProps} />
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
