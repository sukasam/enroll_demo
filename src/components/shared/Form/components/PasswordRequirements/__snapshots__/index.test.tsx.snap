// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordRequirements matches snapshot 1`] = `
<div
  class="passwordRequirements"
>
  <div
    class="passwordLabel"
    data-testid="create_account_password_must"
  >
    create_account_password_must
  </div>
  <ul>
    <li
      class="initial"
      data-testid="create_account_password_requirement_1"
    >
      create_account_password_requirement_1
    </li>
    <li
      class="initial"
      data-testid="create_account_password_requirement_2"
    >
      create_account_password_requirement_2
    </li>
    <li
      class="initial"
      data-testid="create_account_password_requirement_3"
    >
      create_account_password_requirement_3
    </li>
    <li
      class="initial"
      data-testid="create_account_password_requirement_4"
    >
      create_account_password_requirement_4
    </li>
    <li
      class="initial"
      data-testid="create_account_password_requirement_5"
    >
      create_account_password_requirement_5
    </li>
  </ul>
</div>
`;
